src/Poker/Deck.hs

```diff
@@ -0,0 +1,167 @@
+-- |
+-- Module      : Poker.Deck
+-- Description : A deck of cards and its operations.
+-- Copyright   : (c) Ghais Issa, 2021
+module Poker.Deck
+  (
+    Suit(..)
+  , Rank(..)
+  , Card(..)
+  , newCard
+  , randomCard
+  , Deck(..)
+  , stdDeck
+  , shuffleT
+  , shuffle
+  , draw
+  , draw_
+  , draw1
+  , draw1_
+  , remove
+  ) where
+
+import           Data.Bits (shift, (.&.))
+import           Data.List (foldl1', (\\))
+import           Data.Random (RVar)
+import           Data.Random.Distribution.Uniform (uniform)
+import           Data.Random.List (shuffleN, shuffleNT)
+import           Data.Random.RVar (RVarT)
+import           Text.Read (Lexeme (Ident), Read (readPrec), lexP)
+
+data Suit
+  = Club
+  | Diamond
+  | Heart
+  | Spade
+  deriving stock (Read, Show, Eq, Enum, Ord)
+
+data Rank
+  = Two
+  | Three
+  | Four
+  | Five
+  | Six
+  | Seven
+  | Eight
+  | Nine
+  | Ten
+  | Jack
+  | Queen
+  | King
+  | Ace
+  deriving stock (Read, Show, Eq, Enum, Ord)
+
+-- | A card is represented as an int where the lower 4 bits representing the
+-- and the rest represents the rank ranging from 0-12.
+--
+-- Use 'newCard' to construct a new card.
+newtype Card = Card Int deriving stock (Eq, Ord)
+
+-- | Construct a new card.
+newCard :: Rank -> Suit -> Card
+newCard rank suite = Card $ fromEnum rank * 4 + fromEnum suite
+
+
+randomCard :: RVar Card
+randomCard = do
+  suit   <- uniform 0 3
+  rank   <- uniform 0 12
+  return $ newCard (toEnum rank) (toEnum suit)
+
+
+data Deck = Deck !Int ![Card]
```

Comment:
Surprising to keep the size of the deck. Is there any reason not to use vector here, which has O(1) for length, tail, head, which I think is everything we'd need.